### DISCUSSION TITLE
Extending a java class that has a two constructors or has a protected constructor

### DISCUSSION THREAD

**Discussion by michaschulze** [mozilla/rhino#1200] [Q&A] [answered]
I have two in my opinion similar problems. 

1. I would like to extend a class that has two constructors which are overloads. How can I control which one of these constructors gets called? For example, I would like to subclass org.eclipse.swt.widgets.Dialog and Dialog has two constructors. 
2. The other class (org.eclipse.jface.dialogs.Dialog) I would like extend has a protected constructor and if I try to use the JavaAdapter, I will always get an access violation since protected constructors can only be called from a subclass.

Till now, I had no success in subclassing both above case with the help of JavaAdapter. Do you have any idea how this could be reached?

**Comment by p-bakker** [comment]
Might these be of help? https://github.com/mozilla/rhino/issues/627, https://github.com/mozilla/rhino/issues/900#issuecomment-850055726

**Reply by michaschulze** [reply]
The #627 helps to a certain extend for the first problem but not for the second one -- and #900 doesn't help in any way, since calling the super constructor (in my case a protected one) would demand to implement an own constructor in the implementation object. How this can be done is also not clear to me. If there is a way to define a constructor in the implementation object then I assume I could call the protected constructor, which would solve the second problem probably.

**Reply by p-bakker** [reply]
Isn't the whole purpose of a proteced constructor that you cannot call it from a subclass that isn't in the same Package?

When extending a class from JavaScript, I find it quite reasonable that thet subclass is considered to not be in the same package as the super class, hence reasonable that you cannot use a protected constructor

**Reply by michaschulze** [reply]
I fully agree with you about the nature and rules for protected constructors and I already know this, however, I thought using the JavaAdapter it is doing sub-classing internally and if so, a protected constructor should be callable. If I'm wrong here also okay, but then the question would be -- how to do the sub-classing in JavaScript in the implementation object used in the JavaAdapter? As example I would like to extend org.eclipse.jface.dialogs.Dialog which has a protected constructor and is envisioned to be the base class for own Dialogs done by sub-classing it.

**Reply by p-bakker** [reply]
As I see it, the authors of org.eclipse.jface.dialogs.Dialog dont want you to subclass it directly, hence they made the constructor protected and Rhino respects this.

But you can subclass any of the direct known subclasses, like org.eclipse.jface.dialogs.MessageDialog, using the JavaAdapter just fine, as those have public constructors

**Reply by michaschulze** [reply]
IMHO your assumption that the authors of Dialog don't want sub-classing isn't correct, since it is the recommend way of realizing custom dialogs. 

Using another already existing subclass would be probably possible also for my current use-case and I already did this in the past for other things. For example, I implemented the object for the JavaAdapter overwrote protected functions, called the overwritten once by using super$, did the additional stuff, and everything worked as expected. However, what I was always wondering was how this could be reached for the constructor as well; and if I understood this discussion and the statements correctly, Rhino does not supported to call protected constructors using the JavaAdapter in general.

**Reply by tonygermano** [reply]
Just to clarify, is your first problem fixed by passing the arguments to the JavaAdaptor as shown in #627? I see that you commented with the same question on that issue almost two years ago, but you don't show what you tried or the errors you received. I believe the number of arguments that you pass should indicate which constructor you want to call in this case.

**Comment by p-bakker** [comment]
> IMHO your assumption that the authors of Dialog don't want sub-classing isn't correct, since it is the recommend way of realizing custom dialogs.

AFAIK a protected constructor means it can only be called from within the same package, either when subclassing or for instantiation, which makes my assumption correct I'd say

> and if I understood this discussion and the statements correctly, Rhino does not supported to call protected constructors using the JavaAdapter in general

I'm just going off of the info you gave, that you get errors when you try to use the protected constructor when subclassing using the JavaAdapter and off of the info mentioned above about protected constructors. Haven't looked at the actual implementation in Rhino

**Reply by michaschulze** [reply]
> > IMHO your assumption that the authors of Dialog don't want sub-classing isn't correct, since it is the recommend way of realizing custom dialogs.
> 
> AFAIK a protected constructor means it can only be called from within the same package, either when subclassing or for instantiation, which makes my assumption correct I'd say

The Java standard says otherwise [https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.6.2.2](https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.6.2.2) but never mind. 

 
> > and if I understood this discussion and the statements correctly, Rhino does not supported to call protected constructors using the JavaAdapter in general
> 
> I'm just going off of the info you gave, that you get errors when you try to use the protected constructor when subclassing using the JavaAdapter and off of the info mentioned above about protected constructors. Haven't looked at the actual implementation in Rhino

Fully understood and I just thought someone tried this before me or had a similar issue. Thanks for your patience and answers.

**Reply by p-bakker** [reply]
> The Java standard says otherwise https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.6.2.2 but never mind.

Just tried it with plain Java code, creating a class B in package Y that extends class A in package X and trying to call a protected super constructor of A from within the constructor of B and that inded works.

So, if that indeed not works with the JavaAdapter, I guess that just never got implemented. 

Feel free to file a feature request and/or better yet, provide a PR that implements this feature

**Comment by p-bakker** [comment, marked as answer]
Going to mark this question as Answered (as there's not been any more activity):
For 1: answer provided via links to cases that detail how to achieve the desired result
For 2: determined that is not a currently supported feature of the JavaAdapter. Feel free to file an issue or provide a PR
